dotcom-rendering/src/web/components/Picture.tsx

```diff
@@ -140,21 +173,71 @@ export const Picture = ({
 	isMainMedia = false,
 	isLazy = true,
 }: Props) => {
-	const hdpiSources = getSources(role, 'hdpi', imageSources);
-	const mdpiSources = getSources(role, 'mdpi', imageSources);
-	const fallbackSrc = getFallback(role, 'hdpi', imageSources);
-	const sizes = getSizes(role, isMainMedia, format);
+	const hdpiSourceSets = getSourcesForRoleAndResolution(
+		imageSources,
+		role,
+		'hdpi',
+	);
+	const mdpiSourceSets = getSourcesForRoleAndResolution(
+		imageSources,
+		role,
+		'mdpi',
+	);
+	const fallbackSrc = getFallback(
+		hdpiSourceSets.length ? hdpiSourceSets : mdpiSourceSets,
+	);
+
+	// Create an array of breakpoints going from highest to lowest, with 0 as the final option
+	const breakpointSizes: [number, number][] = [
+		...Object.values(breakpoints).reverse(),
+		0,
+	].map((breakpoint) => [
+		breakpoint,
+		getSizeForBreakpoint(breakpoint, role, isMainMedia, format),
+	]);
+	const optimisedBreakPointSizes: [number, number][] =
+		optimiseBreakpointSizes(breakpointSizes);
 
 	return (
 		<picture itemProp="contentUrl">
-			{/* HDPI Source (DPR2) - images in this srcset have `dpr=2&quality=45` in the url */}
-			<source
-				srcSet={hdpiSources}
-				sizes={sizes}
-				media="(-webkit-min-device-pixel-ratio: 1.25), (min-resolution: 120dpi)"
-			/>
-			{/* MDPI Source (DPR1) - images in this srcset have `quality=85` in the url */}
-			<source srcSet={mdpiSources} sizes={sizes} />
+			{format.display === ArticleDisplay.Immersive && isMainMedia ? (
+				<>
+					{portraitImmersiveSource(hdpiSourceSets, true)}
+					{portraitImmersiveSource(mdpiSourceSets, false)}
+				</>
+			) : (
+				''
+			)}
+
+			{optimisedBreakPointSizes.map(([breakpoint, size]) => (
+				<>
+					{/* HDPI Source (DPR2) - images in this srcset have `dpr=2&quality=45` in the url */}
+					{hdpiSourceSets.length ? (
+						<source
+							srcSet={getSourceSetForSize(
+								size,
+								hdpiSourceSets,
+								true,
+							)}
+							sizes={`${breakpoint || size}px`}
```

Comment:
Is there a way to lift this conditional logic up so that `size` is always what you want? It feels very late to have a choice here.

I appreciate this is very 'feely' though and I've yet to fully build a model for the code

Reply:
I found this part a bit confusing when trying to replicate the functionality of Frontend.

On all regular breakpoints, the breakpoint is set as the 'sizes' attribute, however on the `0` breakpoint, Frontend would pass the lowest image size it had instead.

I don't actually think there's any functional difference, whatever the number is the browser is forced to pick an image, though I'd need to test whether it would be okay with receiving '0' as a size or not.